### DISCUSSION TITLE
`Sodium.Core` in `net7.0` work but crashes in `net7.0-maccatalyst`. Why?

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/runtime#79502] [Q&A] [answered]
Hey .NET gurus

Would you possibly know a reason why a console application (<TargetFramework>net7.0</TargetFramework>) works with [`Sodium.Core`]((https://www.nuget.org/packages/Sodium.Core/)) library but when I attempt to call a Sodium.Core function from a vanilla .NET MacCatalyst application  (<TargetFramework>net7.0-maccatalyst</TargetFramework>), my app crashes with `DllNotFoundException` because libsodium native library cannot be found (https://github.com/ektrah/libsodium-core/blob/4cc4f449bf186a825456204909996c9cbafb1175/src/Interop/Interop.Core.cs#L8).

I understand that the question is hard to answer but maybe you can just give me a pointer what might be wrong or how to debug the issue. I have spent days on this but given that I don't know low level details of MacCatalyst & .NET, it's pretty hard for me to understand the issue.

Can it possibly be because I need to bundle somehow the libsodium native library with my Mac Catalyst app?

### Test console app

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <RootNamespace>libsodium_core_console</RootNamespace>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
      <PackageReference Include="Sodium.Core" Version="1.3.1" />
  </ItemGroup>
</Project>
```

```cs
using Sodium;

internal class Program
{
    private static void Main(string[] args)
    {
        Console.WriteLine(SecretAeadAes.IsAvailable);
    }
}
```

### Test MacCatalyst app

1. Create a vanilla MacCatalyst app using VS for MAC
2. Add
    ```xml
    <ItemGroup>
        <PackageReference Include="Sodium.Core" Version="1.3.1" />
    </ItemGroup>
    ```
3. Modify `AppDelegate.cs` as follows:
    ```cs
    namespace libsodium_core_maccatalyst;
    
    using Sodium; // <-- NEW LINE
    
    [Register ("AppDelegate")]
    public class AppDelegate : UIApplicationDelegate {
	    public override UIWindow? Window {
		    get;
		    set;
	    }
    
	    public override bool FinishedLaunching (UIApplication application, NSDictionary launchOptions)
	    {
		    // create a new window instance based on the screen size
		    Window = new UIWindow (UIScreen.MainScreen.Bounds);
    
		    // create a UIViewController with a single UILabel
		    var vc = new UIViewController ();
		    vc.View!.AddSubview (new UILabel (Window!.Frame) {
			    BackgroundColor = UIColor.SystemBackground,
			    TextAlignment = UITextAlignment.Center,
			    Text = SecretAeadAes.IsAvailable.ToString(), // <--- MODIFIED LINE
			    AutoresizingMask = UIViewAutoresizing.All,
		    });
		    Window.RootViewController = vc;
    
		    // make the window visible
		    Window.MakeKeyAndVisible ();
    
		    return true;
	    }
    }
    ```

______
https://github.com/ektrah/libsodium-core/pull/89#issuecomment-1345558725

**Comment by rolfbjarne** [comment, marked as answer]
Could you attach a [binary build log](https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#binary-build-logs) + the exact exception message with stack trace you get?

**Reply by MartyIX** [reply]
This is the exception I get:

<img width="1439" alt="Screenshot 2022-12-12 at 08 42 57" src="https://user-images.githubusercontent.com/203266/206989303-f47c2889-6ddf-4485-81b0-e3d5a0025a94.png">

This is the binlog:

[msbuild.binlog.zip](https://github.com/dotnet/runtime/files/10205623/msbuild.binlog.zip)

**Reply by MartyIX** [reply]
The original application: [libsodium-core-maccatalyst.zip](https://github.com/dotnet/runtime/files/10205638/libsodium-core-maccatalyst.zip)

**Reply by rolfbjarne** [reply]
There's no maccatalyst version of the libsodium dylib here: https://nuget.info/packages/libsodium/1.0.18.4

<img width="215" alt="Screen Shot 2022-12-12 at 09 29 17" src="https://user-images.githubusercontent.com/249268/206997589-b88de4b0-0922-4ad9-8f58-376231422edc.png">

there must be an entry for `maccatalyst-x64` and/or `maccatalyst-arm64`.

**Reply by MartyIX** [reply]
I see. Thanks a lot. Very appreciated. I'll try to fix it!
